File: backend/app/api/v1/customers.py
from __future__ import annotations

from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, status

from app.core.db import AsyncSession, get_db
from app.core.deps import CurrentUser, require_roles
from app.core.responses import success_envelope
from app.models.user import UserRole
from app.schemas.customer import CustomerCreate, CustomerRead, CustomerUpdate
from app.services.customer import CustomerService

router = APIRouter(prefix="/customers", tags=["customers"])

_WRITE_ROLES = (
    UserRole.ADMIN,
    UserRole.INVENTORY_MANAGER,
    UserRole.SALES_REPRESENTATIVE,
)
_READ_ROLES = (
    UserRole.ADMIN,
    UserRole.INVENTORY_MANAGER,
    UserRole.SALES_REPRESENTATIVE,
    UserRole.WAREHOUSE_STAFF,
)


def _svc(session: AsyncSession = Depends(get_db)) -> CustomerService:
    return CustomerService(session)


@router.post(
    "",
    response_model=None,
    status_code=status.HTTP_201_CREATED,
    dependencies=[Depends(require_roles(*_WRITE_ROLES))],
)
async def create_customer(
    body: CustomerCreate,
    current_user: CurrentUser,
    svc: CustomerService = Depends(_svc),
) -> dict[str, object]:
    try:
        customer = await svc.create(body, tenant_id=current_user.tenant_id)
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc)) from exc
    return success_envelope(CustomerRead.model_validate(customer).model_dump(mode="json"))


@router.get(
    "",
    response_model=None,
    dependencies=[Depends(require_roles(*_READ_ROLES))],
)
async def list_customers(
    active_only: bool = Query(False),
    svc: CustomerService = Depends(_svc),
) -> dict[str, object]:
    customers = await svc.list_customers(active_only=active_only)
    return success_envelope(
        [CustomerRead.model_validate(c).model_dump(mode="json") for c in customers]
    )


@router.get(
    "/{customer_id}",
    response_model=None,
    dependencies=[Depends(require_roles(*_READ_ROLES))],
)
async def get_customer(
    customer_id: UUID,
    svc: CustomerService = Depends(_svc),
) -> dict[str, object]:
    try:
        customer = await svc.get(customer_id)
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    return success_envelope(CustomerRead.model_validate(customer).model_dump(mode="json"))


@router.put(
    "/{customer_id}",
    response_model=None,
    dependencies=[Depends(require_roles(*_WRITE_ROLES))],
)
async def update_customer(
    customer_id: UUID,
    body: CustomerUpdate,
    svc: CustomerService = Depends(_svc),
) -> dict[str, object]:
    try:
        customer = await svc.update(customer_id, body)
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    return success_envelope(CustomerRead.model_validate(customer).model_dump(mode="json"))


@router.delete(
    "/{customer_id}",
    response_model=None,
    status_code=status.HTTP_200_OK,
    dependencies=[Depends(require_roles(*_WRITE_ROLES))],
)
async def delete_customer(
    customer_id: UUID,
    svc: CustomerService = Depends(_svc),
) -> dict[str, object]:
    try:
        await svc.delete(customer_id)
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    return success_envelope({"deleted": True})

File: frontend/src/api/customers.ts
import { apiClient, type SuccessEnvelope } from './client'

export interface Customer {
  id: string
  customer_code: string
  business_name: string
  contact_person: string | null
  email: string | null
  phone: string | null
  gst_number: string | null
  drug_license_number: string | null
  address: string | null
  credit_limit: string
  status: string
}

export interface CustomerCreate {
  customer_code: string
  business_name: string
  contact_person?: string
  email?: string
  phone?: string
  gst_number?: string
  drug_license_number?: string
  address?: string
  credit_limit?: string
  status?: string
}

export async function listCustomers(params?: { active_only?: boolean }): Promise<Customer[]> {
  const res = await apiClient.get<SuccessEnvelope<Customer[]>>('/customers', { params })
  return res.data.data
}

export async function getCustomer(id: string): Promise<Customer> {
  const res = await apiClient.get<SuccessEnvelope<Customer>>(`/customers/${id}`)
  return res.data.data
}

export async function createCustomer(data: CustomerCreate): Promise<Customer> {
  const res = await apiClient.post<SuccessEnvelope<Customer>>('/customers', data)
  return res.data.data
}

export async function updateCustomer(id: string, data: Partial<CustomerCreate>): Promise<Customer> {
  const res = await apiClient.put<SuccessEnvelope<Customer>>(`/customers/${id}`, data)
  return res.data.data
}

export async function deleteCustomer(id: string): Promise<void> {
  await apiClient.delete(`/customers/${id}`)
}
